numismatic/feeds/base.py

```diff
@@ -241,15 +238,56 @@ def _make_request(self, api_url, params=None, headers=None, raw=False):
             data = response
         return data
 
-
 @attr.s
 class WebsocketClient(abc.ABC):
     '''Base class for WebsocketClient feeds'''
     # FIXME: Is this really an ABC? What abstractmethods are there?
 
     exchange = None
     websocket_url = None
-    websocket = attr.ib(default=None)
+    _websocket_map = dict()
+    _semaphore_map = dict()
+
+    @property
+    def websocket(self):
+        try:
+            return self._websocket_map[self.websocket_url]
+        except KeyError:
+            return None
+    
+    @property
+    def semaphore(self):
+        if self.websocket_url is None:
+            return None
+        try:
+            return WebsocketClient._semaphore_map[self.websocket_url]
+        except KeyError:
+            semaphore = asyncio.BoundedSemaphore(1)
```

Comment:
I see why the semaphores are necessary but I think this makes the code more complicated than I would like it to be. I want to try and refactor this at some point.